### PR TITLE
TensorGraph Refactor w/ Generators

### PR DESCRIPTION
**Pull request by lilleswing**
See test_tensor_graph and mnist for examples of constructing graphs with the new APIs

### CODE REVIEW

**Comment by peastman** [comment]
Nice!  It looks like the documentation still needs to be updated.

File: deepchem/data/datasets.py [hunk 1]
```diff
@@ -1031,3 +1032,24 @@ def get_label_means(self):
   def get_label_stds(self):
     """Return pandas series of label stds."""
     return self.metadata_df["y_stds"]
+
+
+class Databag(object):
```
Comment (peastman) [inline deepchem/data/datasets.py]: What is this class for?  It needs docstrings.

File: deepchem/utils/evaluate.py [hunk 2]
```diff
@@ -128,3 +129,91 @@ def compute_model_performance(self,
       return multitask_scores
     else:
       return multitask_scores, all_task_scores
+
+
+class GeneratorEvaluator(object):
+  """Class that evaluates a model on a given dataset."""
```
Comment (peastman) [inline deepchem/utils/evaluate.py]: This needs better documentation too.  The description doesn't even make sense, given that no Dataset appears in its API.

File: deepchem/utils/test/test_generator_evaluator.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+    databag = Databag()
+
+    features = Feature(shape=(None, n_features))
+    databag.add_dataset(features, X)
+
+    outputs = []
+    entropies = []
+    labels = []
+    for i in xrange(2):
```
Comment (peastman) [inline deepchem/utils/test/test_generator_evaluator.py]: That should be `range()`.  `xrange()` doesn't exist in Python 3.

File: deepchem/models/tensorgraph/layers.py [hunk 4]
```diff
@@ -165,57 +189,66 @@ class TimeSeriesDense(Layer):
...
 
   def __init__(self, **kwargs):
-    super().__init__(**kwargs)
+    super(LossLayer, self).__init__(**kwargs)
 
-  def __call__(self, *parents):
-    guess, label = parents[0], parents[1]
+  def _create_tensor(self):
+    guess, label = self.in_layers[0], self.in_layers[1]
     self.out_tensor = tf.reduce_mean(
```
Comment (rbharath) [inline deepchem/models/tensorgraph/layers.py]: We should call this `L2LossLayer` since we're doing an L2 operation here.

File: deepchem/models/tensorgraph/tensor_graph.py [hunk 5]
```diff
@@ -170,35 +226,63 @@ def predict_on_batch(self, X, sess=None):
...
 
+    Returns:
+      y_pred: numpy ndarray of shape (n_samples, n_classes*n_tasks)
+    """
     if not self.built:
       self.build()
-    if sess is None:
-      saver = tf.train.Saver()
+
+    def predict_closure(session):
```
Comment (rbharath) [inline deepchem/models/tensorgraph/tensor_graph.py]: Could you add a docstring here? How does `predict_closure` differ from the other predict methods?

File: deepchem/models/tensorgraph/tests/test_tensor_graph.py [hunk 6]
```diff
@@ -1,50 +1,254 @@
...
+    ys = [NumpyDataset(y1), NumpyDataset(y2)]
+
+    databag = Databag()
+
+    features = Feature(shape=(None, n_features))
+    databag.add_dataset(features, X)
+
+    outputs = []
+    entropies = []
+    for i in range(2):
```
Comment (rbharath) [inline deepchem/models/tensorgraph/tests/test_tensor_graph.py]: Our current convention of having a separate placeholder for each label is a little awkward (as @lilleswing has mentioned offline before :-)). In a follow-up PR, let's make a `MultiOutput` layer that makes it easy to build models with many output heads. This should take in only a single `Label` of shape `(None, n_tasks)` or `(None, n_tasks, n_classes)`

File: deepchem/models/tensorgraph/tests/test_tensor_graph.py [hunk 7]
```diff
@@ -1,50 +1,254 @@
...
-
-    g.fit(dataset, nb_epoch=100)
-    g.save()
-    g1 = TensorGraph.load_from_dir('/tmp/tmpss5_ki5_')
-    print(g1)
-    print(g1.predict_on_batch(X))
+  def test_single_task_classifier(self):
+    n_data_points = 200
+    n_features = 10
+    X = np.random.rand(n_data_points, n_features)
```
Comment (rbharath) [inline deepchem/models/tensorgraph/tests/test_tensor_graph.py]: The new API looks really clean! A big step up here.

File: deepchem/models/tensorgraph/tests/test_tensor_graph.py [hunk 8]
```diff
@@ -1,50 +1,254 @@
...
+
+    total_loss = ReduceMean(in_layers=losses)
+
+    tg = dc.models.TensorGraph(learning_rate=0.1)
+    for output in outputs:
+      tg.add_output(output)
+    tg.set_loss(total_loss)
+
+    tg.fit_generator(
+        databag.iterbatches(
```
Comment (rbharath) [inline deepchem/models/tensorgraph/tests/test_tensor_graph.py]: Having to use a `Databag` for a multi-task problem is a little painful. Let's think about ways we can clean this up in follow-on PRs.

File: deepchem/data/datasets.py [hunk 9]
```diff
@@ -1031,3 +1042,45 @@ def get_label_means(self):
...
+  """
+  A utility class to iterate through multiple datasets together.
```
Comment (rbharath) [inline deepchem/data/datasets.py]: This class is a neat helper!

File: examples/tensorgraph/mnist.py [hunk 10]
```diff
@@ -6,82 +6,42 @@
...
+make_image = Reshape(shape=(-1, 28, 28, 1), in_layers=[feature])
 
-conv2d_2 = Conv2d(num_outputs=64)
-tg.add_layer(conv2d_2, parents=[conv2d_1])
+conv2d_1 = Conv2d(num_outputs=32, in_layers=[make_image])
+maxpool_1 = MaxPool(in_layers=[conv2d_1])
 
-flatten = Flatten()
-tg.add_layer(flatten, parents=[conv2d_2])
+conv2d_2 = Conv2d(num_outputs=64, in_layers=[maxpool_1])
```
Comment (rbharath) [inline examples/tensorgraph/mnist.py]: Following Keras's example, let's not use the keyword args in these examples. I like the idea of making examples maximally simple for newbies. More advanced users can decide to use `kwargs` to avoid bugs in mature code.

File: examples/tox21/tox21_tensorgraph_graph_conv.py [hunk 11]
```diff
@@ -18,87 +22,94 @@
...
-
-  degree_slice = Input(shape=(None, 2), dtype=tf.int32)
-  model.add_feature(degree_slice)
 
-  membership = Input(shape=(None,), dtype=tf.int32)
-  model.add_feature(membership)
+def graph_conv_model(batch_size, tasks):
+  model = TensorGraph(
+      model_dir=model_dir, batch_size=batch_size, use_queue=False)
+  atom_features = Feature(shape=(None, 75), name="ATOMFEATURES")
```
Comment (rbharath) [inline examples/tox21/tox21_tensorgraph_graph_conv.py]: Let's remove the key-word args and the explicit names here to make the example maximally simple for newbies.

File: examples/tox21/tox21_tensorgraph_graph_conv.py [hunk 12]
```diff
@@ -6,10 +6,14 @@
 from __future__ import unicode_literals
```
Comment (rbharath) [inline examples/tox21/tox21_tensorgraph_graph_conv.py]: Couple general thoughts about this file:

1. First, amazing work @lilleswing! It's very cool that sophisticated models like these graph-convs can now be built and trained entirely in user-space. This should make algorithmic development in DeepChem much more feasible.
2. Because we've exposed the internals, this code looks messier than our old graph-conv examples. While those did mysterious black-magic under the hood, they had the advantage of looking stylistically straight-forward. Let's brainstorm ways to increase simplicity while keeping algorithms in user-space.

File: examples/tox21/tox21_tensorgraph_graph_conv.py [hunk 13]
```diff
@@ -18,87 +22,94 @@
...
+  gg1 = GraphGather(
+      batch_size=batch_size,
+      activation_fn=tf.nn.tanh,
+      in_layers=[batch_norm3, degree_slice, membership] + deg_adjs,
+      name="GATHER")
 
   costs = []
-  for task in range(num_tasks):
+  labels = []
+  for task in tasks:
```
Comment (rbharath) [inline examples/tox21/tox21_tensorgraph_graph_conv.py]: The new `MultiOutput` layer I proposed would make this part a little simpler.

File: examples/tensorgraph/mnist.py (hunk 10 above)
Comment (mpharrigan) [inline examples/tensorgraph/mnist.py]: I'd argue that kwargs make it much easier to tell what's being passed into a function (for newbies and pros alike)

**Comment by mpharrigan** [comment]
Do you need to introduce a dependency on `six`? Just use `zip`. It will use iterators on python3 and explicitly construct a list on python2 but the observed behavior will be the same. If it's a big problem for users, they can switch to python3. Making code py3 by default with considerations for py2 instead of vice versa will make it easier in the future

File: examples/tensorgraph/mnist.py (hunk 10 above)
Comment (rbharath) [inline examples/tensorgraph/mnist.py]: On checking the actual Keras docs, I see that key-word args are used extensively https://keras.io/getting-started/functional-api-guide/. I retract the comment :-)

File: examples/tox21/tox21_tensorgraph_graph_conv.py (hunk 11 above)
Comment (rbharath) [inline examples/tox21/tox21_tensorgraph_graph_conv.py]: Following comment above, no need to remove the key-word args, but let's get rid of names. If I understand correctly, the names are mainly useful for tensorboard visualization. Let's have a separate Tensorboard example (in a future PR) that walks users through making tensorboard friendly models.

File: deepchem/models/tensorgraph/tensor_graph.py (hunk 5 above)
Comment (lilleswing) [inline deepchem/models/tensorgraph/tensor_graph.py]: I don't need it anymore.  It's a remnant that was needed for a single entry point which session was activated and the context.

File: examples/tox21/tox21_tensorgraph_graph_conv.py (hunk 13 above)
Comment (lilleswing) [inline examples/tox21/tox21_tensorgraph_graph_conv.py]: I'm currently more in favor of a utility function in layers which does this loop under the hood.  Its nice knowing that the results are always of the shape (sample, task, value), and you can get a task value which is always an array by looking at model.outputs.

**Comment by lilleswing** [comment]
@mpharrigan the things we are iterating over are generally going to be DiskDatasets which are purposely sharded and lazy loaded as they won't fit in RAM.  I think it will be safer to stick to using six for now.  Although that is a selfish reason as the Schrodinger deploy of DeepChem will be on python 2.7 and will have to loop over multiple datasets that won't fit in RAM.

**Comment by peastman** [comment]
We discussed that when we moved to Python 3 and specifically decided against that approach.  See #240.

**Comment by lilleswing** [comment]
@peastman sure I can just change it to zip then.  Didn't see that thread earlier.

File: deepchem/models/tensorgraph/tests/test_tensor_graph.py (hunk 8 above)
Comment (lilleswing) [inline deepchem/models/tensorgraph/tests/test_tensor_graph.py]: Yeah lets chat about it.  I personally don't think its too painful,  Its basically mapping Input Tensors to their numpy arrays, but the numpy arrays have all the logic for batching built in to them.

**Comment by rbharath** [comment]
@lilleswing Ok, looks good overall. I think we can go ahead and merge in once Travis passes (The `xrange()` still seems to be in there though, which will break Travis on py3).

**Comment by rbharath** [comment]
It looks like Travis is timing out. Could this be the MNIST test?

**Comment by peastman** [comment]
It still needs the documentation to be updated (docs/using_tensorgraph.md).

**Comment by rbharath** [comment]
Going to go ahead and merge this one in so we can experiment with it (as @peastman points out, we should update docs, but we can do that in follow-on PRs).

LGTM.
